core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/DataProcessor.scala

```diff
@@ -341,20 +313,23 @@ class DataProcessor(
         initBatch(channelId, tuples)
         processInputTuple(Left(inputBatch(currentInputIdx)))
       case EndOfUpstream() =>
-        val currentLink = upstreamLinkStatus.getInputLink(channelId.fromWorkerId)
-        upstreamLinkStatus.markWorkerEOF(channelId.fromWorkerId)
-        if (upstreamLinkStatus.isLinkEOF(currentLink)) {
+        val channel = this.inputGateway.getChannel(channelId)
+        channel.complete()
```

Comment:
Add port to java side, and keep a mapping from channel id to its status. later we may introduce ChannelExecution.

Reply:
done